pkg/querier/frontend/record_rule_substitution.go

```diff
@@ -0,0 +1,331 @@
+	if rr.qrrMap == nil {
+		return []*QueryRangeRequest{r}, nil
+	}
+
+	org, err := user.ExtractOrgID(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	rReplaced, maxModifiedAt, err := rr.qrrMap.ReplaceQueryWithRecordingRule(org, r)
+	if err != nil {
+		return nil, err
+	}
+
+	if rReplaced == nil {
+		return []*QueryRangeRequest{r}, nil
+	}
+
+	modifiedAtMillis := maxModifiedAt.UnixNano() / int64(time.Millisecond/time.Nanosecond)
+	modifiedAtMillis = (modifiedAtMillis / r.Step) * r.Step
+	if modifiedAtMillis > r.End {
+		// No recording rule exist for the query range.
+		return []*QueryRangeRequest{r}, nil
+	}
+	if modifiedAtMillis <= r.Start {
+		// Recording rule exists for entire query range.
+		return []*QueryRangeRequest{rReplaced}, nil
+	}
+
+	// Recording rule exists for partial query range.
+
+	alignedSafetyOffset := (safetyOffset / r.Step) * r.Step
+	r.End = modifiedAtMillis + alignedSafetyOffset
+
+	rReplaced.Start = r.End + r.Step
+	if rReplaced.Start >= rReplaced.End {
+		// Very less evaulations using the recording rule.
+		// This is also possible because of adding 'alignedSafetyOffset'.
+		return []*QueryRangeRequest{r}, nil
+	}
+
+	return []*QueryRangeRequest{r, rReplaced}, nil
+}
+
+// generaliseQuery returns the query by formatting it with PromQL printer.
+// This will generalise the query structure, and also check for errors in query.
+func generaliseQuery(q string) (string, error) {
+	expr, err := promql.ParseExpr(q)
+	if err != nil {
+		return "", err
+	}
+	return expr.String(), nil
+}
+
+// RecordRuleSubstitutionConfig gives thread safe access to
+// query to recording rule map of all organisations.
+type RecordRuleSubstitutionConfig struct {
+	qrrMap map[string][]QueryToRecordingRuleMap // (org id) -> its query-recording rule maps.
+	mtx    sync.RWMutex
+}
+
+// LoadFromFile clears the previous config and loads the query to recording rule
+// config from a file.
+func (oqr *RecordRuleSubstitutionConfig) LoadFromFile(filename string) error {
+	b, err := ioutil.ReadFile(filename)
+	if err != nil {
+		oqr.qrrMap = nil
+		return err
+	}
+	return oqr.LoadFromBytes(b)
+}
+
+// LoadFromBytes clears the previous config and loads the query to recording rule
+// config from bytes.
+func (oqr *RecordRuleSubstitutionConfig) LoadFromBytes(b []byte) (err error) {
+	oqr.mtx.Lock()
+	defer func() {
+		if err != nil {
+			oqr.qrrMap = nil
+		}
+		oqr.mtx.Unlock()
+	}()
+
+	var parsedFile recordRuleSubstitutionConfigFile
+	if err = yaml.UnmarshalStrict(b, &parsedFile); err != nil {
+		return err
+	}
+
+	// Creating the query to recording rule map from the file.
+	oqr.qrrMap = make(map[string][]QueryToRecordingRuleMap)
+	for _, o := range parsedFile.Orgs {
+		for i := range o.QueryToRecordingRuleMap {
+			o.QueryToRecordingRuleMap[i].Query, err = generaliseQuery(o.QueryToRecordingRuleMap[i].Query)
+			if err != nil {
+				return err
+			}
+		}
+		oqr.qrrMap[o.OrgID] = append(oqr.qrrMap[o.OrgID], o.QueryToRecordingRuleMap...)
+	}
```

Comment:
Could we store the recording rules in a format which means we don't need to do this translation?